### PR TITLE
[WIP] #1987, #807, #2446, #2598 Changes needed for CKEditor5 upgrade

### PR DESCRIPTION
**Pull request by joeparsons**
## Description
WIP changes needed for upgrading to CKEditor5.

- Upgrades `az_standard` text format and editor configs to CKEditor5
- Upgrades `basic` text format and editor configs to CKEditor5
- Installs CKEditor5 module instead of CKEditor (4) module by default on new site installs
- Adds DB update to install CKEditor5 module on existing sites
- Updates icons for custom embed buttons to SVGs (needed for CKEditor5)
- Adds CKEditor5 stylesheets to AZ Barrio theme

Questions we still need to consider:
- Is it going to be safe for people to apply these config changes via Config Distro if they have customized the text formats or editors in any way?
- When (if ever) do we want to automate the uninstallation of the contrib ckeditor module?
- Should we remove the remaining CKeditor 4 specific things from Quickstart now or later? (AZ Barrio ckeditor styles, `az_media_ckeditor_styles_alter()`, etc)

See validation messages that came up when switching editor/text_format configs to CKeditor5 here: https://github.com/az-digital/az_quickstart/issues/1987#issuecomment-1640724617

## Related issues

- #807
- #1987
- #2446
- #2598

## How to test

Verify that editor works with new and existing content at least as well as it did with CKEditor 4.

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [x] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: az_quickstart.info.yml [hunk 1]
```diff
@@ -33,6 +33,7 @@ install:
   - drupal:block
   - drupal:breakpoint
   - ckeditor:ckeditor
+  - drupal:ckeditor5
```
Comment (joeparsons) [inline az_quickstart.info.yml:36]:
```suggestion
  - drupal:ckeditor5
```

**Comment by Carris-Toupin** [comment]
Here are the issues I have found:

1. ul do not look WYSIWYG in the Editing text box. They are pushed to the left and hard to see at all.
2. ol does not number/letter correctly in the View mode
3. Tables:

**Comment by trackleft** [comment]
Here is what I see with UL, OL, and Tables
https://arizona.box.com/s/89ytqke6a3714ia846eyikukl3dupjox

**Comment by joeparsons** [comment]
I was able to reproduce issues with unordered lists, at least, with existing (demo) content and after saving new content and then editing the field.

**Comment by Carris-Toupin** [comment]
1. 1.	UL: does not look correct in the WYSIWYG text box in Edit. You cannot see the markers (they are too far to the left) and the indented levels do not show.
a.	I cannot seem to use the Styles for UL or the triangle bullets.
2.	OL: looks correct in the WYSIWYG text box in Edit, but has the wrong numbers/letter arrangement in View Mode.
3.	Tables:
a.	Table Caption is on top of the table in the WYSIWYG text box, but is under the table in View Mode. This is misleading.
b.	I cannot get vertical lines, might be me
c.	I had no horizontal lines, then I did. No idea why. The lines act weird. 
4.	Paragraph, you must click paragraph to go back to ‘normal’ text.
5.	Not sure how to use Block Quote Footer. I have not found when the focus is on it so it can be used.
This is what I have so far. There most likely will be more.

**Comment by trackleft** [comment]
I will look into a reset to override the CKEditor reset seen here https://ckeditor.com/docs/ckeditor5/latest/installation/advanced/content-styles.html#the-full-list-of-content-styles

**Comment by trackleft** [comment]
I am going to move all our custom library settings from az_barrio to az_core

In order to prevent multiple instances of arizona-bootstrap loaded on any give page within the admin UI where CKEditor is generally loaded, I am planning on using the libraries API from Drupal Core.

Hopefully, I'll be able to set a stub library in az_barrio that adds az_core/arizona-bootsrap as a dependency so we don't have any issues with subthemes or non-quickstart custom modules that depend on `az_barrio/arizona-bootstrap`

**Comment by joeparsons** [comment]
## Decisions made during 2023-09-06 AZ Digital workshop meeting

- We need to migrate the Arizona Bootstrap Drupal CSS/JS libraries and their associated configuration settings from the AZ Barrio theme to the AZ Core module so that the dynamic library definitions (which are modified by `hook_libraries_info_alter()`) can load consistently in the admin theme and in the front-end theme.  This will fix the problems with different versions of Arizona Bootstrap getting loaded in the admin theme when editing content.
- Instead of removing the Arizona Bootstrap libraries and settings from the AZ Barrio theme right away, we will instead deprecate them as of AZQS 2.8.0 and plan to remove them in AZQS 2.9.0 in case any site owners rely on these in their custom modules or subthemes.  AZQS 2.8.x will include 2 Arizona Bootstrap Drupal libraries: one defined by az_core and one defined by AZ Barrio (deprecated).
- We need to remove the dependency on the `az_barrio/global-styling` library that exists in various `az_paragraphs` module defined libraries as styles from that library are being applied to content in the CKeditor and making things look weird.
- We will likely also need to add more CKeditor specific styles / overrides in `az_barrio/css/ckeditor5.css` to address any remaining issues with styles not rendering the same within CKeditor as on the front end.
- @trackleft is going to create a separate / follow-up issue about streamlining the way we manage / maintain the value of the latest Arizona Bootstrap release in AZQS.

**Comment by joeparsons** [comment]
## Tasks remaining

- [ ] Formalize plan for removing libraries and settings from AZ Barrio theme in 2.9.0 and how to deal with differences between module settings and and theme settings (needs separate ticket)
- [ ] Disable form elements in theme settings form and add deprecation notice
- [ ] Update all libraries that depend on AZ Barrio defined libraries to use AZ Core libraries
- [ ] Update anything that uses theme settings to use module settings
- [ ] Document steps for updating custom module/custom subtheme that relies on libraries/settings
- [ ] Fix coding standards issues

File: .github/workflows/arizona-bootstrap-release.yml [hunk 2]
```diff
@@ -22,9 +22,6 @@ jobs:
           git checkout -b 'az-bootstrap-${{ github.event.client_payload.version }}'
           jq --indent 4 '.require."az-digital/arizona-bootstrap" |= "${{ github.event.client_payload.version }}"' composer.json > composer.json.new
           mv composer.json.new composer.json
-          cat themes/custom/az_barrio/includes/common.inc | sed "s/^define('AZ_BOOTSTRAP_STABLE_VERSION'.*/define('AZ_BOOTSTRAP_STABLE_VERSION', '${{ github.event.client_payload.version }}');/g" > common.inc.new
-          mv common.inc.new themes/custom/az_barrio/includes/common.inc
-          git add composer.json themes/custom/az_barrio/includes/common.inc
```
Comment (trackleft) [inline .github/workflows/arizona-bootstrap-release.yml:27]: Add all of this back

File: themes/custom/az_barrio/theme-settings.php [hunk 3]
```diff
@@ -12,8 +12,6 @@
 
 use Drupal\Core\Form\FormStateInterface;
 use Drupal\Core\File\Exception\FileException;
-use Drupal\Core\Link;
-use Drupal\Core\Url;
 use Drupal\Core\StreamWrapper\StreamWrapperManager;
```
Comment (trackleft) [inline themes/custom/az_barrio/theme-settings.php:15]: Add back

File: themes/custom/az_barrio/theme-settings.php [hunk 4]
```diff
@@ -92,155 +90,9 @@ function az_barrio_form_system_theme_settings_alter(&$form, FormStateInterface $
 
   // Fonts and Icons.
   unset($form['fonts']['fonts']['bootstrap_barrio_google_fonts']);
-  $form['fonts']['fonts']['az_barrio_font'] = [
```
Comment (trackleft) [inline themes/custom/az_barrio/theme-settings.php:95]: Disable form elements instead of removing.
Also add deprecation notes.

**Comment by trackleft** [comment]
While attempting to write a ticket/issue about streamlining the way we manage / maintain the value of the latest Arizona Bootstrap release in AZQS i really couldn't think of any clear advantages to doing this, so I propose we just drop that issue.

File: themes/custom/az_barrio/theme-settings.php [hunk 5]
```diff
@@ -92,6 +92,7 @@ function az_barrio_form_system_theme_settings_alter(&$form, FormStateInterface $
 
   // Fonts and Icons.
   unset($form['fonts']['fonts']['bootstrap_barrio_google_fonts']);
+  $form['fonts']['fonts']['#description'] = '<div class="alert alert-info messages info">' . t('<strong>NOTE:</strong> With the advent of CKEditor5, Quickstart\'s settings for choosing how your site includes fonts have moved. These theme settings have been deprecated in Quickstart 2.8.0 and will be removed in Quickstart 2.9.0 in favor of Quickstart Core configuration.') . '</div>';
```
Comment (joeparsons) [inline themes/custom/az_barrio/theme-settings.php:95]: It'd probably be good to the Quickstart core settings form from here, too.

File: themes/custom/az_barrio/includes/common.inc [hunk 6]
```diff
@@ -5,12 +5,8 @@
  * common.inc
  *
  * Contains common functionality for the entire theme.
+ * @deprecated Use modules/custom/az_core/includes/common.inc instead.
```
Comment (trackleft) [inline themes/custom/az_barrio/includes/common.inc:8]:
```suggestion
 * @deprecated in 2.8.0 and is removed from 2.9.0. Use modules/custom/az_core/includes/common.inc instead.
```

File: themes/custom/az_barrio/includes/common.inc [hunk 7]
```diff
@@ -5,12 +5,8 @@
  * common.inc
  *
  * Contains common functionality for the entire theme.
+ * @deprecated in 2.8.0 and is removed from 2.9.0. Use modules/custom/az_core/includes/common.inc instead.
  */
```
Comment (trackleft) [inline themes/custom/az_barrio/includes/common.inc:9]:
```suggestion
 * @see https://github.com/az-digital/az_quickstart/pull/2596
 */
```

File: modules/custom/az_core/src/Form/QuickstartCoreSettingsForm.php [hunk 8]
```diff
@@ -2,13 +2,16 @@
 
 namespace Drupal\az_core\Form;
 
+use Drupal\az_core\Utility\AzLibraryVersionChecker;
```
Comment (trackleft) [inline modules/custom/az_core/src/Form/QuickstartCoreSettingsForm.php:5]:
```suggestion
```

File: modules/custom/az_core/src/Form/QuickstartCoreSettingsForm.php [hunk 9]
```diff
@@ -212,6 +218,156 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
+                  'attributes' => [
+                    'target' => '_blank',
+                  ],
+                ]
+            )
+          )->toString(),
+        ],
+
+        ),
+      'local' => t('Use local copy of AZ Icons packaged with AZ Barrio (%stableversion).', ['%stableversion' => AZ_ICONS_STABLE_VERSION]),
```
Comment (trackleft) [inline modules/custom/az_core/src/Form/QuickstartCoreSettingsForm.php:366]:
```suggestion
        'local' => t('Use local copy of AZ Icons packaged with AZ Barrio (%stableversion).', ['%stableversion' => AZ_ICONS_STABLE_VERSION]),
```

File: themes/custom/az_barrio/theme-settings.php (hunk 5 above)
Comment (trackleft) [inline themes/custom/az_barrio/theme-settings.php:95]:
```suggestion
  $form['fonts']['fonts']['#description'] = '<div class="alert alert-info messages info">' . t('<strong>NOTE:</strong> With the advent of CKEditor5, Quickstart"'"s settings for choosing how your site includes fonts have moved. These theme settings have been deprecated in Quickstart 2.8.0 and will be removed in Quickstart 2.9.0 in favor of Quickstart Core configuration.') . '</div>';
```

File: themes/custom/az_barrio/theme-settings.php [hunk 10]
```diff
@@ -124,6 +127,7 @@ function az_barrio_form_system_theme_settings_alter(&$form, FormStateInterface $
     '#title' => t('Icons'),
     '#collapsible' => TRUE,
     '#collapsed' => FALSE,
+    '#description' => '<div class="alert alert-info messages info">' . t('<strong>NOTE:</strong> With the advent of CKEditor5, Quickstart\'s settings for choosing how your site includes icons have moved. These theme settings have been deprecated in Quickstart 2.8.0 and will be removed in Quickstart 2.9.0 in favor of Quickstart Core configuration.') . '</div>',
```
Comment (trackleft) [inline themes/custom/az_barrio/theme-settings.php:130]:
```suggestion
    '#description' => '<div class="alert alert-info messages info">' . t('<strong>NOTE:</strong> With the advent of CKEditor5, Quickstart"'"s settings for choosing how your site includes icons have moved. These theme settings have been deprecated in Quickstart 2.8.0 and will be removed in Quickstart 2.9.0 in favor of Quickstart Core configuration.') . '</div>',
```

File: themes/custom/az_barrio/theme-settings.php [hunk 11]
```diff
@@ -201,6 +208,7 @@ function az_barrio_form_system_theme_settings_alter(&$form, FormStateInterface $
     '#title' => t('Arizona Bootstrap'),
     '#group' => 'bootstrap',
     '#weight' => -9,
+    '#description' => '<div class="alert alert-info messages info">' . t('<strong>NOTE:</strong> With the advent of CKEditor5, Quickstart\'s settings for choosing how your site includes Arizona Bootstrap have moved. These theme settings have been deprecated in Quickstart 2.8.0 and will be removed in Quickstart 2.9.0 in favor of Quickstart Core configuration.') . '</div>',
```
Comment (trackleft) [inline themes/custom/az_barrio/theme-settings.php:211]:
```suggestion
    '#description' => '<div class="alert alert-info messages info">' . t('<strong>NOTE:</strong> With the advent of CKEditor5, Quickstart"'"s settings for choosing how your site includes Arizona Bootstrap have moved. These theme settings have been deprecated in Quickstart 2.8.0 and will be removed in Quickstart 2.9.0 in favor of Quickstart Core configuration.') . '</div>',
```

File: themes/custom/az_barrio/theme-settings.php [hunk 12]
```diff
@@ -92,6 +92,7 @@ function az_barrio_form_system_theme_settings_alter(&$form, FormStateInterface $
 
   // Fonts and Icons.
   unset($form['fonts']['fonts']['bootstrap_barrio_google_fonts']);
+  $form['fonts']['fonts']['#description'] = '<div class="alert alert-info messages info">' . t('<strong>NOTE:</strong> With the advent of CKEditor5, Quickstart"'"s settings for choosing how your site includes fonts have moved. These theme settings have been deprecated in Quickstart 2.8.0 and will be removed in Quickstart 2.9.0 in favor of Quickstart Core configuration.') . '</div>';
```
Comment (trackleft) [inline themes/custom/az_barrio/theme-settings.php:95]:
```suggestion
  $form['fonts']['fonts']['#description'] = '<div class="alert alert-info messages info">' . t('<strong>NOTE:</strong> With the advent of CKEditor5, Quickstart"'"s settings for choosing how your site includes fonts have moved. These theme settings have been deprecated in Quickstart 2.8.0 and will be removed in Quickstart 2.9.0 in favor of @quickstart-core-link
          , [
          '@quickstart-core-link' => Link::fromTextAndUrl(
            'Quickstart Core configuration', Url::fromRoute(
                'az_core.az_settings', [
                  'attributes' => [
                    'target' => '_blank',
                  ],
                ]
            )
          )->toString(),
        ],
.') . '</div>';
```

File: modules/custom/az_core/az_core.install [hunk 13]
```diff
@@ -12,3 +12,20 @@ function az_core_update_last_removed() {
   // Remove updates added before 2.6.0.
   return 920501;
 }
+
+/**
+ * Implements hook_update().
```
Comment (joeparsons) [inline modules/custom/az_core/az_core.install:17]:
```suggestion
 * Migrate config for AZ Bootstrap, AZ Icons, and Font from AZ Barrio theme settings to AZ Core module config.
```

**Comment by trackleft** [comment]
Replaced by #2769

**Comment by joeparsons** [comment]
Like @trackleft said, we decided to go with #2769 instead.

**Pull request closed without merging**